Windows HostProcess Container Documentation (#28413)
kubernetes/website: content/en/docs/tasks/configure-pod-container/create-hostprocess-pod.md

- As of v1.22 HostProcess pods can only run HostProcess containers. This is a current limitation
 of the Windows OS; non-privileged Windows containers cannot share a vNIC with the host IP namespace.
- HostProcess containers run as a process on the host and do not have any degree of 
isolation other than resource constraints imposed on the HostProcess user account. Neither 
filesystem or Hyper-V isolation are supported for HostProcess containers.
- Volume mounts are supported and are mounted under the container volume. See [Volume Mounts](./create-hostprocess-pod#volume-mounts)
- A limited set of host user accounts are available for HostProcess containers by default. See [Choosing a User Account](./create-hostprocess-pod#choosing-a-user-account).
- Resource limits (disk, memory, cpu count) are supported in the same fashion as processes 
on the host.
- Both Named pipe mounts and Unix domain sockets are **not** currently supported and should instead be accessed via their 
path on the host (e.g. \\\\.\\pipe\\\*)

 ## {{% heading "prerequisites" %}}
# HostProcess Documentation

> To enable HostProcess containers while in Alpha you need to pass the following feature gate flag to kubelet and kube-apiserver. See [Features Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview)documentation for more details.

Reviewer: ```suggestion
To enable HostProcess containers while in Alpha you need to pass the following feature gate flag to **kubelet** and **kube-apiserver**. See [Features Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview) documentation for more details.
```

---
title: Create a Windows HostProcess Pod
content_type: task
weight: 20

> ---

Reviewer 2: ```suggestion
min-kubernetes-server-version: 1.22
---
```

log onto each Window node. HostProcess containers can run as any user that is 
available on the host or is in the domain of the host machine, allowing administrators 
to restrict resource access through user permissions. While neither filesystem or process 
isolation are supported, a new volume is created on the host upon starting the container 
to give it a clean and consolidated workspace. HostProcess containers can also be built on 
top of existing Windows base images and do not inherit the same 
[compatibility requirements](https://docs.microsoft.com/virtualization/windowscontainers/deploy-containers/version-compatibility) as Windows server containers.

### When should I use a Windows HostProcess container?

- When you need to perform tasks which require the networking namespace of the host. 
HostProcess containers have access to the host's network interfaces and IP.
- You need access to resources on the host such as the filesystem, event logs, etc.
- Installation of specific devices drivers or services
- Consolidation of administrative tasks and security policies. This reduces the degree of 
privileges needed by Windows nodes.

> ### Important Notes

Reviewer 2: ```suggestion
## Limitations
```
?

- As of v1.22 HostProcess pods can only run HostProcess containers. This is a current limitation
 of the Windows OS; non-privileged Windows containers cannot share a vNIC with the host IP namespace.
- HostProcess containers run as a process on the host and do not have any degree of 
isolation other than resource constraints imposed on the HostProcess user account. Neither 
filesystem or Hyper-V isolation are supported for HostProcess containers.
- Volume mounts are supported and are mounted under the container volume. See [Volume Mounts](./create-hostprocess-pod#volume-mounts)
- A limited set of host user accounts are available for HostProcess containers by default. See [Choosing a User Account](./create-hostprocess-pod#choosing-a-user-account).
- Resource limits (disk, memory, cpu count) are supported in the same fashion as processes 
on the host.
- Both Named pipe mounts and Unix domain sockets are **not** currently supported and should instead be accessed via their 
path on the host (e.g. \\\\.\\pipe\\\*)

> ## {{% heading "prerequisites" %}}

Reviewer 2: This should be listed before “Important Notes”.

> 

Reviewer 2: ```suggestion

{{% version-check %}}

```
(goes with `min-kubernetes-server-version` in the front matter)

To enable HostProcess containers while in Alpha you need to pass the following feature gate flag to **kubelet** and **kube-apiserver**. 

> See [Features Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview) documentation for more details.

Reviewer 2: nit:
```suggestion
See [Features Gates](/docs/reference/command-line-tools-reference/feature-gates/#overview) documentation for more details.
```

```
--feature-gates=WindowsHostProcessContainers=true
```

> You can use the latest version of Containerd (v1.5.4+) with the following settings using the containerd v2 configuration.  

Reviewer 2: Will this advice stay current? It's better to reword so that this sentence stays true with little maintenance needed.

Once the containerd shim is built you can replace the file in your contianerd installation.  For example if you followed the instructions
 to [install containerd](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd) replace the
  `containerd-shim-runhcs-v1.exe` is installed at `$Env:ProgramFiles\containerd` with the newly built shim.

We are working to improve this process by enabling HostProcess support directly in 
[containerd via the CRI api](https://github.com/containerd/containerd/pull/5131) as well as shipping a version of the containerd shim
 for Windows with support direclty via containerd releases.

## Creating a HostProcess Pod

### Security Policy Config
Enabling a Windows HostProcess pod simply requires setting the right configurations in the pod security configuration. 
Many of the policies in the [Pod Security Standards](/docs/concepts/security/pod-security-standards) do not apply to 
HostProcess containers (and Windows in general) due to architectural differences between Windows and Linux. As such, 

> here are the policy configurations required to enable HostProcess pods:

Reviewer 3: This confuses me a bit.  I don't know how this fits into the https://kubernetes.io/docs/concepts/security/pod-security-standards/#policy-types.  Would these be additional fields in the `Privileged ` type?

Author: Updated to make this more clear.

          <li><code>"NT AUTHORITY\\Local service - Win32 Apps"</code></li>
          <li><code>"NT AUTHORITY\\SYSTEM"</code></li>
				</ul>
			</td>
		</tr>
    <tr>
			<td style="white-space: nowrap"><a href="/docs/concepts/security/pod-security-standards">runAsNonRoot</a></td>
			<td>
				<p>HostProcess containers have privileged access to the host so the runAsNonRoot field cannot be set to true.</p>
				<p><strong>Allowed Values</strong></p>
				<ul>
          <li>Undefined/Nil</li>
					<li><code>false</code></li>
				</ul>
			</td>
		</tr>
	</tbody>
</table>

### Example Manifest

```yaml
spec:
  securityContext:
    windowsOptions:
      hostProcess: true
      runAsUserName: "NT AUTHORITY\\Local service"
  hostNetwork: true
  containers:
  - name: test
    image: image1:latest
    command:
      - ping
      - -t
      - 127.0.0.1
  nodeSelector:
    "kubernetes.io/os": windows
```

## Volume Mounts

HostProcess containers support the ability to mount volumes within the container volume space. Applications running inside
 the container can access volume mounts directly via relative or absolute paths. An environment variable `$CONTAINER_SANDBOX_MOUNT_POINT`

>   is set upon container creation and provides the absolute host path to the container volume. Relative paths are based upon

Reviewer 3: spaces at the begining of these lines are off

Reviewer 3: and through out this document

   the `Pod.containers.volumeMounts.mountPath` configuration.

Example: To access service account tokens the following path structures are supported within the container:

> `\var\run\secrets\kubernetes.io\serviceaccount\`

Reviewer 3: I am confused by this one... I don't think this will work, the forward slash in front would cause it to look up the `c:\var\...`

`$CONTAINER_SANDBOX_MOUNT_POINT\var\run\secrets\kubernetes.io\serviceaccount\`

## Choosing a User Account

> Currently HostProcess containers only support the ability to run as one of four supported Windows service accounts:

Author: This is reasonable, it's likely that it will be expanded to include password or custom local accounts but this is under investigation.

- **[LocalService](https://docs.microsoft.com/en-us/windows/win32/services/localservice-account)**
- **[NetworkService](https://docs.microsoft.com/en-us/windows/win32/services/networkservice-account)**
- **[LocalSystem](https://docs.microsoft.com/en-us/windows/win32/services/localsystem-account)**

> - **[LocalServiceAccount](https://docs.microsoft.com/en-us/windows/win32/services/localservice-account)**

Reviewer 3: this looks like a duplicate of the first one in the list

> Operators should use these accounts to restrict access to system resources (e.g. files, registry, named pipes, WMI, etc.)

Reviewer 3: how? By adding additional acls to these accounts?

---
title: Create a Windows HostProcess Pod
content_type: task
weight: 20
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.22" state="alpha" >}}

Windows HostProcess containers enable you to run containerized 
workloads on a Windows host. These containers operate as 
normal processes but have access to the host network namespace, 
storage, and devices when given the appropriate user privileges. 

> HostProcess containers can be used to deploy CNIs, 

Reviewer 2: nit:
```suggestion
HostProcess containers can be used to deploy network plugins,
```

components to Windows nodes without the need for dedicated proxies or 
the direct installation of host services.

Administrative tasks such as installation of security patches, event 
log collection, and more can be performed without requiring cluster operators to 
log onto each Window node. HostProcess containers can run as any user that is 
available on the host or is in the domain of the host machine, allowing administrators 
to restrict resource access through user permissions. While neither filesystem or process 
isolation are supported, a new volume is created on the host upon starting the container 
to give it a clean and consolidated workspace. HostProcess containers can also be built on 
top of existing Windows base images and do not inherit the same 
[compatibility requirements](https://docs.microsoft.com/virtualization/windowscontainers/deploy-containers/version-compatibility) 
as Windows server containers, meaning that the version of the base images does not need 
to match that of the host.

### When should I use a Windows HostProcess container?

- When you need to perform tasks which require the networking namespace of the host. 

> HostProcess containers have access to the host's network interfaces and IP.

Reviewer 2: nit:
```suggestion
HostProcess containers have access to the host's network interfaces and IP addresses.
```
Last time I used Windows server I think it enabled both IPv4 and IPv6 by default, so it's normal to find a host with more than one IP address.

- You need access to resources on the host such as the filesystem, event logs, etc.

> - Installation of specific devices drivers or services

Reviewer 2: ```suggestion
- Installation of specific device drivers or Windows services.
```

- Consolidation of administrative tasks and security policies. This reduces the degree of 
privileges needed by Windows nodes.

### Important Notes

> - HostProcess containers will **only run when using version 1.5.4 (or higher) of the containerd [container runtime](https://kubernetes.io/docs/setup/production-environment/container-runtimes/)**.

Reviewer 2: ```suggestion
- HostProcess containers require version 1.5.4 or higher of the containerd {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}.
```

- Volume mounts are supported and are mounted under the container volume. 
See [Volume Mounts](./create-hostprocess-pod#volume-mounts)
- A limited set of host user accounts are available for HostProcess containers by default. 
See [Choosing a User Account](./create-hostprocess-pod#choosing-a-user-account).
- Resource limits (disk, memory, cpu count) are supported in the same fashion as processes 
on the host.
- Both Named pipe mounts and Unix domain sockets are **not** currently supported and should instead 
be accessed via their path on the host (e.g. \\\\.\\pipe\\\*)

## {{% heading "prerequisites" %}}

To enable HostProcess containers while in Alpha you need to pass the following feature gate flag to 
**kubelet** and **kube-apiserver**. 
See [Features Gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview) 
documentation for more details.

```
--feature-gates=WindowsHostProcessContainers=true
```

You can use the latest version of Containerd (v1.5.4+) with the following settings using the containerd 
v2 configuration. Add these annotations to any runtime configurations were you wish to enable the 

> HostProcess Container feature.

Reviewer 2: nit:
```suggestion
HostProcess container feature.
```

```
[plugins]
  [plugins."io.containerd.grpc.v1.cri"]
    [plugins."io.containerd.grpc.v1.cri".containerd]
      [plugins."io.containerd.grpc.v1.cri".containerd.default_runtime]
        container_annotations = ["microsoft.com/hostprocess-container"]
        pod_annotations = ["microsoft.com/hostprocess-container"]
    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process]
        container_annotations = ["microsoft.com/hostprocess-container"]
        pod_annotations = ["microsoft.com/hostprocess-container"]
```

The current versions of containerd ship with a version of hcsshim that does not have support. 
You will need to build a version of hcsshim from the main branch following the 
[instructions in hcsshim](https://github.com/Microsoft/hcsshim/#containerd-shim). 
Once the containerd shim is built you can replace the file in your contianerd installation. 
For example if you followed the instructions to 

> [install containerd](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd) 

Reviewer 2: nit:
```suggestion
[install containerd](/docs/setup/production-environment/container-runtimes/#containerd)
```

			<td>
				<p>Windows pods offer the ability to run <a href="/docs/tasks/configure-pod-container/create-hostprocess-pod">
        HostProcess containers</a> which enables privileged access to the Windows node. </p>
				<p><strong>Allowed Values</strong></p>
				<ul>
					<li><code>true</code></li>
				</ul>
			</td>
		</tr>
		<tr>
			<td style="white-space: nowrap"><a href="/docs/concepts/security/pod-security-standards">Host Networking</a></td>
			<td>
				<p>Will be in host network by default initially. Support 
				to set network to a different compartment may be desirable in 
				the future.</p>
				<p><strong>Allowed Values</strong></p>
				<ul>
					<li><code>true</code></li>
				</ul>
			</td>
		</tr>
    <tr>
			<td style="white-space: nowrap"><a href="/docs/tasks/configure-pod-container/configure-runasusername/">runAsUsername</a></td>
			<td>
				<p>Specification of which user the HostProcess container should run as is required for the pod spec.</p>
				<p><strong>Allowed Values</strong></p>
				<ul>
          <li><code>"NT AUTHORITY\\SYSTEM"</code></li>
					<li><code>"NT AUTHORITY\\Local service"</code></li>

>           <li><code>"NT AUTHORITY\\NetworkService"</code></li>

Reviewer 2: I think it'd be clearer to write these unquoted and with just one backslash. In the API, the double backslash is part of the encoding (eg JSON) - at least, that's my understanding. The manifest below shows how to write that in YAML.

We could add a comment in the YAML to explain why two backslashes are needed there.

				</ul>
			</td>
		</tr>
    <tr>
			<td style="white-space: nowrap"><a href="/docs/concepts/security/pod-security-standards">runAsNonRoot</a></td>
			<td>

> 				<p>HostProcess containers have privileged access to the host so the runAsNonRoot field cannot be set to true.</p>

Reviewer 2: ```suggestion
				<p>Because HostProcess containers have privileged access to the host, the <tt>runAsNonRoot</tt> field cannot be set to true.</p>
```

				<p><strong>Allowed Values</strong></p>
				<ul>
          <li>Undefined/Nil</li>
					<li><code>false</code></li>
				</ul>
			</td>
		</tr>
	</tbody>
</table>

### Example Manifest

```yaml
spec:
  securityContext:
    windowsOptions:
      hostProcess: true
      runAsUserName: "NT AUTHORITY\\Local service"
  hostNetwork: true
  containers:
  - name: test
    image: image1:latest
    command:
      - ping
      - -t
      - 127.0.0.1
  nodeSelector:
    "kubernetes.io/os": windows
```

## Volume Mounts

HostProcess containers support the ability to mount volumes within the container volume space. 
Applications running inside the container can access volume mounts directly via relative or 
absolute paths. An environment variable `$CONTAINER_SANDBOX_MOUNT_POINT` is set upon container 
creation and provides the absolute host path to the container volume. Relative paths are based 
upon the `Pod.containers.volumeMounts.mountPath` configuration.

> Example: To access service account tokens the following path structures are supported within the container:

Reviewer 2: ```suggestion
### Example {#volume-mount-example}

To access service account tokens the following path structures are supported within the container:
```

I'm not sure if this _is_ an example though. These seem to be the only two paths you might use? I'm not sure.

> `.\var\run\secrets\kubernetes.io\serviceaccount\`

Reviewer 2: Could you use a UNC path?

Reviewer 3: In this case the security account token is mapped in by kubelet so won't be on a UNC path.  

In general the system accounts should be able to use UNC paths if they have access. Volume mounting them into the container might not make sense and I think the guidance would be the same as the guidance for named pipes.  @dcantah  to confirm

Reviewer 4: A relative path like this would only work if the working directory was untouched. The default working directory is where the volume is mounted, but a different one could be set in the image or through cri etc. so this wouldn't work in that case. Good to call out.

Reviewer 4: We had a little race on the comment :P, but yea the guidance would be to just use the path on the host. As Kubelet maps this in explicitly I think that's why we call this scenario out

`$CONTAINER_SANDBOX_MOUNT_POINT\var\run\secrets\kubernetes.io\serviceaccount\`

## Choosing a User Account

Currently HostProcess containers only support the ability to run as one of three supported Windows service accounts:

- **[LocalSystem](https://docs.microsoft.com/en-us/windows/win32/services/localsystem-account)**
- **[LocalService](https://docs.microsoft.com/en-us/windows/win32/services/localservice-account)**
- **[NetworkService](https://docs.microsoft.com/en-us/windows/win32/services/networkservice-account)**

> Operators should use these accounts to limit the degree of privileges made available to the HostProcess container to prevent accidental damage to the host. The LocalSystem account has the highest level of privilege of the three and should be used only if absolutely necessary. It is recommended that operators follow the principle of least privilege and use the LocalService account when possible.

Reviewer 2: ```suggestion
You should select an appropriate Windows service account for each HostProcess
container, aiming to limit the degree of privileges so as to avoid accidental (or even
malicious) damage to the host. The LocalSystem servuce account has the highest level
of privilege of the three and should be used only if absolutely necessary. Where possible,
use the LocalService service account as it is the least privileged of the three options.
```

 of the Windows OS; non-privileged Windows containers cannot share a vNIC with the host IP namespace.
- HostProcess containers run as a process on the host and do not have any degree of 
isolation other than resource constraints imposed on the HostProcess user account. Neither 
filesystem or Hyper-V isolation are supported for HostProcess containers.
- Volume mounts are supported and are mounted under the container volume. See [Volume Mounts](./create-hostprocess-pod#volume-mounts)
- A limited set of host user accounts are available for HostProcess containers by default. See [Choosing a User Account](./create-hostprocess-pod#choosing-a-user-account).
- Resource limits (disk, memory, cpu count) are supported in the same fashion as processes 
on the host.
- Both Named pipe mounts and Unix domain sockets are **not** currently supported and should instead be accessed via their 
path on the host (e.g. \\\\.\\pipe\\\*)

 ## {{% heading "prerequisites" %}}
[todo] Version requirements
[todo] Enabling the feature
- flags that need to be set 
- 

- list of steps needed to build everything

## Creating HostProcess containers

> [todo] Deploying HostProcess containers

Reviewer 2: Here's an example of work that still needs doing (I'm afraid)

it might be OK to cut down the documentation rather than try to fill this out. Otherwise, if the docs aren't ready for review, I think the enhancement could become at risk for v1.22.

top of existing Windows base images and do not inherit the same 
[compatibility requirements](https://docs.microsoft.com/virtualization/windowscontainers/deploy-containers/version-compatibility) 
as Windows server containers, meaning that the version of the base images does not need 
to match that of the host. HostProcess containers also support 

> [volume mounts](./create-hostprocess-pod#volume-mounts) within the container volume.

Reviewer 2: Please make this a page link:
```suggestion
[volume mounts](#volume-mounts) within the container volume.
```

- HostProcess containers require version 1.5.4 or higher of the containerd {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}.
- As of v1.22 HostProcess pods can only contain HostProcess containers. This is a current limitation 
of the Windows OS; non-privileged Windows containers cannot share a vNIC with the host IP namespace.
- HostProcess containers run as a process on the host and do not have any degree of 
isolation other than resource constraints imposed on the HostProcess user account. Neither 
filesystem or Hyper-V isolation are supported for HostProcess containers.
- Volume mounts are supported and are mounted under the container volume. 
See [Volume Mounts](./create-hostprocess-pod#volume-mounts)
- A limited set of host user accounts are available for HostProcess containers by default. 
See [Choosing a User Account](./create-hostprocess-pod#choosing-a-user-account).
- Resource limits (disk, memory, cpu count) are supported in the same fashion as processes 
on the host.
- Both Named pipe mounts and Unix domain sockets are **not** currently supported and should instead 
be accessed via their path on the host (e.g. \\\\.\\pipe\\\*)

> ## Creating a HostProcess Security Policy Config

Reviewer 2: ```suggestion
## Creating a HostProcess security policy configuration
```

I'm not sure what you create when you follow this step. What is it that the reader creates?

Author: Good point, I was thinking of it as 'what to do for HostProcess when creating your configuration'. I'll reword it to 'HostProcess Pod configuration requirements'
